backend/remote-state/kubernetes/client.go

```diff
@@ -0,0 +1,342 @@
+	lockInfoKey    = "lockInfo"
+)
+
+type RemoteClient struct {
+	k8sClient  dynamic.ResourceInterface
+	namespace  string
+	nameSuffix string
+	workspace  string
+}
+
+func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
+	sName, err := c.createSecretName()
+	if err != nil {
+		return nil, err
+	}
+	secret, err := c.k8sClient.Get(sName, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	sData := getSecretData(secret)
+
+	stateRaw, ok := sData[terraState]
+	if !ok {
+		// The secret exists but there is no state in it
+		return nil, nil
+	}
+
+	stateRawString := stateRaw.(string)
+
+	state, err := uncompressState(stateRawString)
+	if err != nil {
+		return nil, err
+	}
+
+	md5 := md5.Sum(state)
+
+	p := &remote.Payload{
+		Data: state,
+		MD5:  md5[:],
+	}
+	return p, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	sName, err := c.createSecretName()
+	if err != nil {
+		return err
+	}
+
+	payload, err := compressState(data)
+	if err != nil {
+		return err
+	}
+
+	secret, err := c.getSecret(sName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		return fmt.Errorf("secret does not exist so lock is not held, secret name: %v err: %v", sName, err)
+	}
+
+	lockInfo, err := c.getLockInfo(secret)
+	if err != nil {
+		return err
+	}
+
+	setTerraState(secret, payload)
+
+	secret, err = c.k8sClient.Update(secret, metav1.UpdateOptions{})
+	if err != nil {
+		lockErr := &state.LockError{
+			Info: lockInfo,
+			Err:  fmt.Errorf("error updating the state: %v", err),
+		}
+		return lockErr
+	}
+
+	return err
+}
+
+// Delete the state secret
+func (c *RemoteClient) Delete() error {
+	sName, err := c.createSecretName()
+	if err != nil {
+		return err
+	}
+
+	err = c.deleteSecret(sName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	sName, err := c.createSecretName()
+	if err != nil {
+		return "", err
+	}
+
+	lockInfo := info.Marshal()
+
+	secret, err := c.getSecret(sName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return "", err
+		}
+
+		// The secret doesn't exist yet, create it with the lock
+		sData := make(map[string][]byte)
+		sData[lockInfoKey] = lockInfo
+
+		label := map[string]string{
+			terraState:     "true",
+			terraKey:       c.nameSuffix,
+			terraWorkspace: c.workspace,
+		}
+		secret = &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"metadata": metav1.ObjectMeta{
+					Name:        sName,
+					Namespace:   c.namespace,
+					Labels:      label,
+					Annotations: map[string]string{"encoding": "gzip"},
+				},
+				"data": sData,
+			},
+		}
+
+		_, err = c.k8sClient.Create(secret, metav1.CreateOptions{})
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return "", err
+			}
+			// The secret was created between the get and create, grab it and keep going
```

Comment:
I'm a bit confused by this.
It's a known fact that the `Create` operation in K8s' client-go is not atomic. And so it makes sense to check whether "our" Create call actually produced the object we're later manipulating.
So what confuses me is why is the next line just reading the already existing object (supposedly produced outside of this client) and moving on using that?

Reply:
If we attempt to create the secret and it already exists, it just needs to be fetched and then fall back to the same flow as if we found the secret the first time we tried to get it.